### PR TITLE
Add APIs for adding and removing tags from rooms

### PR DESCRIPTION
**Pull request by NegativeMjark**
Tested by matrix-org/sytest#34
Documented by matrix-org/matrix-doc#132

TODO:

* [x] Add an event stream for private_user_data
* [x] Hook the event stream for private user_data to the notifier.
* [x] Add tags to the v1 /initialSync and /events.
* [x] Add tags to the v2 /sync

### CODE REVIEW

File: synapse/rest/client/v2_alpha/tags.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        if user_id != auth_user.to_string():
+            raise AuthError(403, "Cannot add tags for other users.")
+
+        max_id = yield self.store.remove_tag_from_room(user_id, room_id, tag)
+
+        yield self.notifier.on_new_event(
+            "private_user_data_key", max_id, users=[user_id]
+        )
+
+        # TODO: poke the notifier.
```
Comment (illicitonion) [inline synapse/rest/client/v2_alpha/tags.py]: Does the line above this not do so?

File: synapse/storage/tags.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        self.get_tags_for_user.invalidate((user_id,))
+
+        result = yield self._private_user_data_id_gen.get_max_token(self)
+        defer.returnValue(result)
+
+    @defer.inlineCallbacks
+    def remove_tag_from_room(self, user_id, room_id, tag):
+        """Remove a tag from a room for a user.
+        Returns:
+            A deffered that completes once the tag has been removed
```
Comment (illicitonion) [inline synapse/storage/tags.py]: s/deffered/deferred/

File: synapse/handlers/private_user_data.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+class PrivateUserDataEventSource(object):
+    def __init__(self, hs):
+        self.store = hs.get_datastore()
+
+    def get_current_key(self, direction='f'):
+        return self.store.get_max_private_user_data_stream_id()
+
+    @defer.inlineCallbacks
+    def get_new_events_for_user(self, user, from_key, limit):
```
Comment (illicitonion) [inline synapse/handlers/private_user_data.py]: FYI I'm changing this method signature on a separate branch; I imagine you'll get this in first, though

File: synapse/rest/client/v2_alpha/tags.py (hunk 1 above)
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/tags.py]: Fixed 06986e4

File: synapse/storage/tags.py (hunk 2 above)
Comment (NegativeMjark) [inline synapse/storage/tags.py]: Fixed 5897e77

**Comment by illicitonion** [comment]
LGTM

**Pull request merged**